File: spark-scala/src/main/scala/qr/spark/partition/MyPartitioner.scala
package qr.spark.partition

import org.apache.spark.{Partition, Partitioner}

/**
 *
 * 要实现自定义分区器，需要继承org.apache.spark.Partitioner类，并实现下面三个方法。
（1）numPartitions: Int:返回创建出来的分区数。
（2）getPartition(key: Any): Int:返回给定键的分区编号（0到numPartitions-1）。
（3）equals():Java 判断相等性的标准方法。这个方法的实现非常重要，Spark需要用这个方法来检查你的分区器对象是否和其他分区器实例相同，这样Spark才可以判断两个RDD的分区方式是否相同。

 * @param partitions 分区数
 */
class MyPartitioner(partitions:Int) extends Partitioner{
  override def numPartitions: Int = partitions

  override def getPartition(key: Any): Int = {

    key match {
      case _:String => 0
      case i:Int => i%numPartitions
      case _ => 0
    }
  }
}

File: hadoop/src/main/java/com/qr/daemon/kafka/partitioner/MyPartitioner.java
package com.qr.daemon.kafka.partitioner;

import org.apache.kafka.clients.producer.Partitioner;
import org.apache.kafka.common.Cluster;

import java.util.Map;

/**
 * 自定义分区器
 *
 */
public class MyPartitioner implements Partitioner {

    /**
     * 分区方法
     * @param topic topic
     * @param key key
     * @param keyBytes key字节码
     * @param value value
     * @param valueBytes value字节码
     * @param cluster 分区个数
     * @return 分区号
     */
    @Override
    public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {

        //1、 获取key
        String keySrt = key.toString();
        //2、 创建分区返回结果
        int parNum;
        //3、 计算key的hash值
        int keyStrHash = keySrt.hashCode();
        //4、 获取topic个数
        Integer partitionNumber = cluster.partitionCountForTopic(topic);
        //5、 计算分区号
        parNum = Math.abs(keyStrHash) % partitionNumber;
        //6、 返回分区号
        return parNum;
    }

    /**
     * 关闭资源
     */
    @Override
    public void close() {

    }

    /**
     * 配置方法
     * @param map map
     */
    @Override
    public void configure(Map<String, ?> map) {

    }
}
